File: 387. First Unique Character in a String.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    int firstUniqChar(string s) {
        vector<int> alphabet(26, 0); // record letter times
        for (int i = 0; i < s.length(); ++i) {
            alphabet[s[i] - 'a'] += 1;
        }
        for (int j = 0; j < s.length(); ++j) {
            if (alphabet[s[j] - 'a'] == 1)
                return j;
        }
        return -1;
    }
};

File: 387. First Unique Character in a String.py
# 387. First Unique Character in a String
# Time:  O(N)
# Space: O(N)

from collections import Counter

# Counter to get frequency, and iterate once
class Solution:
    def firstUniqChar(self, s):
        """
        :type s: str
        :rtype: int
        """
        letter_cnt = Counter(s)
        for i, c in enumerate(s):
            if letter_cnt[c] == 1:
                return i
        return -1
        